___

<a href='https://mainacademy.ua/'> <img src='https://mainacademy.ua/wp-content/uploads/2019/03/logo-main.png' alt = 'Хмм, щось з інтернетом'/></a>
___

# Module 5: Data analysis with NumPy and Pandas

## Lab work 5.2



#### Мета: 

* навчитися працювати із бібліотекою Pandas в Python.

### Завдання:

In [1]:
import pandas as pd

**Вивести версію та конфігурацію бібліотеки:**

In [2]:
print(pd.__version__)

print(pd.get_option('display.colheader_justify'))
print(pd.get_option('display.encoding'))
print(pd.get_option('display.max_columns'))
print(pd.get_option('display.max_colwidth'))
print(pd.get_option('display.max_rows'))
print(pd.get_option('display.min_rows'))

1.2.2
right
UTF-8
20
50
60
10


In [3]:
df = pd.read_csv('Ecommerce Purchases')

**Скільки рядків і стовпців в наборі даних:**

In [4]:
print(f'В наборі даних', df.shape[0], 'рядків та', df.shape[1], 'стовпців')

В наборі даних 10000 рядків та 14 стовпців


**Перевірити, чи в наборі даних містяться порожні значення:**

In [5]:
N = df.isnull().sum().reset_index()
print(f'Кількість порожніх значень -',N[0].sum())

Кількість порожніх значень - 0


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Address           10000 non-null  object 
 1   Lot               10000 non-null  object 
 2   AM or PM          10000 non-null  object 
 3   Browser Info      10000 non-null  object 
 4   Company           10000 non-null  object 
 5   Credit Card       10000 non-null  int64  
 6   CC Exp Date       10000 non-null  object 
 7   CC Security Code  10000 non-null  int64  
 8   CC Provider       10000 non-null  object 
 9   Email             10000 non-null  object 
 10  Job               10000 non-null  object 
 11  IP Address        10000 non-null  object 
 12  Language          10000 non-null  object 
 13  Purchase Price    10000 non-null  float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB


**Яка середня ціна закупки (Purchase Price):**

In [7]:
print(f'Середня ціна закупки (Purchase Price) -',df['Purchase Price'].mean())

Середня ціна закупки (Purchase Price) - 50.347302


In [8]:
df.describe()

Credit Card  CC Security Code  Purchase Price
count  1.000000e+04      10000.000000    10000.000000
mean   2.341374e+15        907.217800       50.347302
std    2.256103e+15       1589.693035       29.015836
min    6.040186e+10          0.000000        0.000000
25%    3.056322e+13        280.000000       25.150000
50%    8.699942e+14        548.000000       50.505000
75%    4.492298e+15        816.000000       75.770000
max    6.012000e+15       9993.000000       99.990000

**Скільки людей користуються англійською мовою "en" на веб-сайті:**

In [9]:
L = df[df['Language'] == 'en']
print(f'Кількість людей, які користуються англійською мовою -', L.shape[0])

Кількість людей, які користуються англійською мовою - 1098


In [10]:
df['Language'].value_counts().reset_index()

index  Language
0    ru      1155
1    de      1155
2    el      1137
3    pt      1118
4    en      1098
5    fr      1097
6    es      1095
7    it      1086
8    zh      1059

**Скільки людей має посаду «Lawyer»?**

In [11]:
def Law(string):
    return True if "Lawyer" in string else False

print(f'Посаду "Lawyer" займають', df[df['Job'].map(Law)].shape[0], 'людей')

Посаду "Lawyer" займають 30 людей


In [12]:
df['Job'].value_counts().reset_index()

index  Job
0        Interior and spatial designer   31
1                               Lawyer   30
2                    Social researcher   28
3                   Purchasing manager   27
4    Research officer, political party   27
..                                 ...  ...
618                    Music therapist    7
619                 Editor, film/video    6
620                            Actuary    6
621                 Investment analyst    5
622        Information systems manager    5

[623 rows x 2 columns]

**Скільки людей зробило покупку вранці та скільки людей зробило покупку після обіду?**

In [13]:
L = df[df['AM or PM'] == 'AM']
print(f'Кількість людей, які зробили покупку до обіду -', L.shape[0])
print(f'Кількість людей, які зробили покупку після обіду -', df.shape[0]-L.shape[0])

Кількість людей, які зробили покупку до обіду - 4932
Кількість людей, які зробили покупку після обіду - 5068


In [14]:
df['AM or PM'].value_counts().reset_index(0)

index  AM or PM
0    PM      5068
1    AM      4932

**Які 5 найпоширеніших назв вакансій?**

In [15]:
df['Job'].value_counts().reset_index()[0:5]

index  Job
0      Interior and spatial designer   31
1                             Lawyer   30
2                  Social researcher   28
3                 Purchasing manager   27
4  Research officer, political party   27

**Хтось здійснив покупку, яка надійшла від Lot: "90 Wt", та якою була ціна придбання для цієї транзакції?**

In [16]:
df.loc[df.Lot == '90 WT', 'Purchase Price'].values[0]

75.1

**Яка електронна адреса особи з таким номером кредитної картки: 4926535242672853 ?**

In [17]:
df.loc[df['Credit Card'] == 4926535242672853, 'Email'].values[0]

'bondellen@williams-garza.com'

**Скільки людей використовує American Express  і здійснили покупку на суму понад 100 доларів?**

In [18]:
df.loc[((df['CC Provider'] == 'American Express') & (df['Purchase Price'] > 100))]

Empty DataFrame
Columns: [Address, Lot, AM or PM, Browser Info, Company, Credit Card, CC Exp Date, CC Security Code, CC Provider, Email, Job, IP Address, Language, Purchase Price]
Index: []

In [19]:
df.loc[((df['CC Provider'] == 'American Express') & (df['Purchase Price'] > 100))].shape[0]

0

In [20]:
df['Credit Card'].nunique()

10000

Жодна людина не здійснила покупку на суму понад 100 доларів. max('Purchase Price') = 99.99

**Скільки людей мають кредитну картку, термін дії якої закінчується в 2025 році?**

In [21]:
df['CC Exp Year'] = df['CC Exp Date'].map(lambda x: x[3:])
df.loc[((df['CC Exp Year'] == '25'))].shape[0]

1033

**Які найкращі 5 найпопулярніших постачальників / хостів електронної пошти (наприклад, gmail.com, yahoo.com тощо ...).**

In [22]:
df['Email domain'] = df['Email'].map(lambda x: x.split('@')[1])
df['Email domain'].value_counts().reset_index().head()

index  Email domain
0   hotmail.com          1638
1     yahoo.com          1616
2     gmail.com          1605
3     smith.com            42
4  williams.com            37

**Виведіть зведену таблицю по браузерах(Browser Info), посаді(Job), та кількості транзакцій :**

In [23]:
df['Browser'] = df['Browser Info'].map(lambda x: str(x).split('/')[0])
df['Job1'] = df['Job'].map(lambda x: str(x).split(',')[0])
pd.pivot_table(df, index = ['Browser', 'Job1'], values = ['Purchase Price'], aggfunc=[len])

len
                                Purchase Price
Browser Job1                                  
Mozilla Academic librarian                10.0
        Accommodation manager             11.0
        Accountant                        48.0
        Accounting technician             10.0
        Actor                             12.0
...                                        ...
Opera   Water quality scientist            5.0
        Web designer                       7.0
        Wellsite geologist                 2.0
        Writer                             1.0
        Youth worker                       3.0

[993 rows x 1 columns]

In [24]:
df['Browser'] = df['Browser Info'].map(lambda x: str(x).split('/')[0])
df.groupby(['Browser', 'Job']).count()

Address  Lot  AM or PM  \
Browser Job                                                        
Mozilla Academic librarian                     10   10        10   
        Accommodation manager                  11   11        11   
        Accountant, chartered                  14   14        14   
        Accountant, chartered certified        13   13        13   
        Accountant, chartered management       10   10        10   
...                                           ...  ...       ...   
Opera   Water quality scientist                 5    5         5   
        Web designer                            7    7         7   
        Wellsite geologist                      2    2         2   
        Writer                                  1    1         1   
        Youth worker                            3    3         3   

                                          Browser Info  Company  Credit Card  \
Browser Job                                                                    
Mozilla Academic librarian                          10       10           10   
        Accommodation manager                       11       11           11   
        Accountant, chartered                       14       14           14   
        Accountant, chartered certified             13       13           13   
        Accountant, chartered management            10       10           10   
...                                                ...      ...          ...   
Opera   Water quality scientist                      5        5            5   
        Web designer                                 7        7            7   
        Wellsite geologist                           2        2            2   
        Writer                                       1        1            1   
        Youth worker                                 3        3            3   

                                          CC Exp Date  CC Security Code  \
Browser Job                                                               
Mozilla Academic librarian                         10                10   
        Accommodation manager                      11                11   
        Accountant, chartered                      14                14   
        Accountant, chartered certified            13                13   
        Accountant, chartered management           10                10   
...                                               ...               ...   
Opera   Water quality scientist                     5                 5   
        Web designer                                7                 7   
        Wellsite geologist                          2                 2   
        Writer                                      1                 1   
        Youth worker                                3                 3   

                                          CC Provider  Email  IP Address  \
Browser Job                                                                
Mozilla Academic librarian                         10     10          10   
        Accommodation manager                      11     11          11   
        Accountant, chartered                      14     14          14   
        Accountant, chartered certified            13     13          13   
        Accountant, chartered management           10     10          10   
...                                               ...    ...         ...   
Opera   Water quality scientist                     5      5           5   
        Web designer                                7      7           7   
        Wellsite geologist                          2      2           2   
        Writer                                      1      1           1   
        Youth worker                                3      3           3   

                                          Language  Purchase Price  \
Browser Job                                                          
Mozilla Aca

**Створіть нову колонку "Actual price", яка утворюється із "Purchase Price" та націнки за принципом:**
- якщо "Purchase Price" > 50, націнка 20%
- якщо "Purchase Price" > 100, націнка 10%
- в інших випадках націнка 30%
   

In [25]:
df.loc[df['Purchase Price'] > 100, 'Actual price'] = df.loc[df['Purchase Price'] > 100, 'Purchase Price']*1.1

df.loc[(df['Purchase Price'] > 50) & (df['Purchase Price'] <= 100), 'Actual price'] = \
df.loc[(df['Purchase Price'] > 50) & (df['Purchase Price'] <= 100), 'Purchase Price']*1.2


df.loc[df['Purchase Price'] <= 50, 'Actual price'] = df.loc[df['Purchase Price'] <= 50, 'Purchase Price']*1.3

df[['Purchase Price', 'Actual price']]

Purchase Price  Actual price
0              98.14       117.768
1              70.73        84.876
2               0.95         1.235
3              78.04        93.648
4              77.82        93.384
...              ...           ...
9995           82.21        98.652
9996           25.63        33.319
9997           83.98       100.776
9998           38.84        50.492
9999           67.59        81.108

[10000 rows x 2 columns]

**Зробіть рангування набору даних по "Language" та "Actual price" в порядку спадання ціни.**

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rank.html

In [26]:
df[['Language','Actual price']].sort_values(by = 'Actual price', ascending = False).rank(ascending = True)

Language  Actual price
2092    1724.0        9999.5
7807     578.0        9999.5
9856     578.0        9998.0
7604    1724.0        9997.0
3844    1724.0        9996.0
...        ...           ...
2783     578.0           4.5
8656    8364.0           4.5
6863    3938.0           3.0
2876    6125.5           1.5
5487    9471.0           1.5

[10000 rows x 2 columns]

**Колонку "Language" (категоріальна змінна) "закодуйте", тобто утворити індикаторні колоник. В наборі не повинна залишитися колонка "Language".**

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rank.html

In [27]:
df = pd.get_dummies(df, columns = ['Language'])

In [28]:
df

Address    Lot AM or PM  \
0     16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1     9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                      Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   
3                 7780 Julia Fords\nNew Stacy, WA 45798  36 vm       PM   
4     23012 Munoz Drive Suite 337\nNew Cynthia, TX 5...  20 IE       AM   
...                                                 ...    ...      ...   
9995      966 Castaneda Locks\nWest Juliafurt, CO 96415  92 XI       PM   
9996  832 Curtis Dam Suite 785\nNorth Edwardburgh, T...  41 JY       AM   
9997              Unit 4434 Box 6343\nDPO AE 28026-0283  74 Zh       AM   
9998               0096 English Rest\nRoystad, IA 12457  74 cL       PM   
9999     40674 Barrett Stravenue\nGrimesville, WI 79682  64 Hr       AM   

                                           Browser Info  \
0     Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1     Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2     Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3     Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4     Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   
...                                                 ...   
9995  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/5352 ...   
9996  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
9997  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_7...   
9998  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_8;...   
9999  Mozilla/5.0 (X11; Linux i686; rv:1.9.5.20) Gec...   

                              Company       Credit Card CC Exp Date  \
0                     Martinez-Herman  6011929061123406       02/20   
1     Fletcher, Richards and Whitaker  3337758169645356       11/18   
2          Simpson, Williams and Pham      675957666125       08/19   
3     Williams, Marshall and Buchanan  6011578504430710       02/24   
4           Brown, Watson and Andrews  6011456623207998       10/25   
...                               ...               ...         ...   
9995                    Randall-Sloan   342945015358701       03/22   
9996         Hale, Collins and Wilson   210033169205009       07/25   
9997                     Anderson Ltd  6011539787356311       05/21   
9998                         Cook Inc   180003348082930       11/17   
9999                       Greene Inc  4139972901927273       02/19   

      CC Security Code                  CC Provider  \
0                  900                 JCB 16 digit   
1                  561                   Mastercard   
2                  699                 JCB 16 digit   
3                  384                     Discover   
4                  678  Diners Club / Carte Blanche   
...                ...                          ...   
9995               838                 JCB 15 digit   
9996               207                 JCB 16 digit   
9997                 1                VISA 16 digit   
9998               987             American Express   
9999               302                 JCB 15 digit   

                               Email  ... Actual price Language_de  \
0                  pdunlap@yahoo.com  ...      117.768           0   
1                 anthony41@reed.com  ...       84.876           0   
2     amymiller@morales-harrison.com  ...        1.235           1   
3        brent16@olson-robinson.info  ...       93.648           0   
4        christopherwright@gmail.com  ...       93.384           0   
...                              ...  ...          ...         ...   
9995          iscott@wade-garner.com  ...       98.652           0   
9996              mary85@hotmail.com  ...       33.319           0   
9997               tyler16@gmail.com  ...      100.776           0   
9998         elizabethmoore@reid.net  ...       50.492           0   
9999           rachelford@vaughn.com  ...       81.108           0   

      Language_el Language_en Language_es Language_fr Language_it  \
0             

**Кінцевий набір даних збережіть у файл з розширенням csv.**

In [29]:
df.to_csv('Lab work 5.2.csv', index = False)